src/main/scala/firrtl/transforms/InlineBooleanExpressions.scala

```diff
@@ -0,0 +1,268 @@
+  * input c : UInt<1>
+  * input d : UInt<1>
+  * output y : UInt<1>
+  * output z : UInt<1>
+  * node _y = and(a, b)
+  * node _z = and(a, b)
+  * y <= and(_y, c)
+  * z <= and(_z, d)
+  * // Replacement of _y and _z is performed because the expressions are assigned to two different nodes}}}
+  *
+  * Transformed to:
+  *
+  * {{{circuit Top :
+  * module Top :
+  * input a : UInt<1>
+  * input b : UInt<1>
+  * input c : UInt<1>
+  * input d : UInt<1>
+  * output y : UInt<1>
+  * output z : UInt<1>
+  * node _y = and(a, b)
+  * node _z = and(a, b)
+  * y <= and(_y, c)
+  * z <= and(_z, d)}}}
+  *
+  * No replacement of _x is performed because _x has a fanout of 2
+  *
+  * Original:
+  *
+  * {{{circuit Top :
+  * module Top :
+  * input a : UInt<1>
+  * input b : UInt<1>
+  * input c : UInt<1>
+  * input d : UInt<1>
+  * output y : UInt<1>
+  * output z : UInt<1>
+  * node _x = and(a, b)
+  * y <= and(_x, c)
+  * z <= and(_x, d)}}}
+  *
+  * Transformed to:
+  *
+  * {{{circuit Top :
+  * module Top :
+  * input a : UInt<1>
+  * input b : UInt<1>
+  * input c : UInt<1>
+  * input d : UInt<1>
+  * output y : UInt<1>
+  * output z : UInt<1>
+  * node _y = and(a, b)
+  * y <= and(_y, c)
+  * z <= and(_z, d)}}}
+  */
+class InlineBooleanExpressions extends Transform {
+  def inputForm = LowForm
+  def outputForm = LowForm
+
+  import InlineBooleanExpressions._
+
+  def makeModuleInfo(c: Circuit): (Map[String, Map[String, String]], Seq[String], InstanceGraph, Map[String, DefModule]) = {
+    val iGraph = (new InstanceGraph(c))
+    val childInstances = iGraph.getChildrenInstances.mapValues(_.map(inst => inst.name -> inst.module).toMap).toMap
+
+    // Order from leaf modules to root so that any module driving an output
+    // with a constant will be visible to modules that instantiate it
+    // TODO Generating order as we execute constant propagation on each module would be faster
+    val x = iGraph.moduleOrder.reverse
+
+    val moduleNameTraversalOrder = x.map { m => m.name }
+
+    val modNameModules = x.map { m => (m.name -> m) }.toMap
+    (childInstances, moduleNameTraversalOrder, iGraph, modNameModules)
+  }
+
+  /** Build a [[Netlist]] from a Module's connections and Nodes
+    * This assumes [[firrtl.LowForm LowForm]]
+    *
+    * @param mod [[firrtl.ir.Module Module]] from which to build a [[Netlist]]
+    * @return [[Netlist]] of the module's connections and nodes
+    */
+  def buildNetlist(mod: DefModule): Netlist = {
+    val netlist = new Netlist()
+    def onStmt(stmt: Statement): Statement = {
+      println(s"stmt = ${stmt.serialize}")
+
+      stmt.map(onStmt) match {
+        case Connect(_, lhs, rhs) if (isTmpVar(lhs.serialize)) => netlist(lhs.serialize) = rhs
+        case DefNode(_, nname, rhs) if (isTmpVar(nname)) => netlist(nname) = rhs
+        case _ => // Do nothing
+      }
+      stmt
+    }
+    mod.map(onStmt)
+    netlist
+  }
+
+  def buildSymTab(mod: DefModule): Map[WrappedExpression, Int] = {
```

Comment:
I think rewording this as `buildReferenceCounter` would be more clear. Also, it would be great if you wrote scaladoc for this as well, since it is a public member of a class.